From 3c22e5dd612d79a5518843b37e9809883a1e0bd5 Mon Sep 17 00:00:00 2001
From: Marek Marczykowski <marmarek@invisiblethingslab.com>
Date: Mon, 6 May 2013 03:29:53 +0200
Subject: [PATCH] minios: ignore close(0)
Organization: Invisible Things Lab
Cc: Marek Marczykowski <marmarek@invisiblethingslab.com>

Stdout is special console, initialized by
init_console()->xencons_ring_init() with info from start_info struct.
Especially some settings like dev->backend, dev->ring_ref are not
initialize there. Ignore close() call on this, because free_consfront()
tries to access those not initialized fields (at least
strlen(dev->backend)). We don't want close our log channel anyway.

This will fix GFP during exit() call (newlib version), which tries to
close all file descriptors.

Signed-off-by: Marek Marczykowski <marmarek@invisiblethingslab.com>
---
 extras/mini-os/lib/sys.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/extras/mini-os/lib/sys.c b/extras/mini-os/lib/sys.c
index a62e039..663506a 100644
--- a/extras/mini-os/lib/sys.c
+++ b/extras/mini-os/lib/sys.c
@@ -420,6 +420,8 @@ int fsync(int fd) {
 int close(int fd)
 {
     printk("close(%d)\n", fd);
+    if (fd == 0)
+        return 0;
     switch (files[fd].type) {
         default:
 	    files[fd].type = FTYPE_NONE;
-- 
1.8.1.4

